src/v/kafka/client/client.cc

```diff
@@ -350,7 +350,42 @@ ss::future<fetch_response> client::fetch_partition(
                            return _brokers.find(leader);
                        })
                        .then([&tp, &build_request](shared_broker_t&& b) {
-                           return b->dispatch(build_request(tp));
+                           return b->dispatch(build_request(tp))
+                             .then([b, &tp](fetch_response res) {
+                                 if (res.data.error_code != error_code{}) {
+                                     return ss::make_exception_future<
+                                       fetch_response>(broker_error(
+                                       b->id(), res.data.error_code));
+                                 }
+
+                                 if (res.data.topics.empty()) {
+                                     return ss::make_exception_future<
+                                       fetch_response>(partition_error(
+                                       tp, error_code::unknown_server_error));
+                                 }
+
+                                 if (res.data.topics.size() == 1) {
+                                     auto& topic = res.data.topics[0];
+                                     if (topic.partitions.empty()) {
+                                         return ss::make_exception_future<
+                                           fetch_response>(partition_error(
+                                           tp,
+                                           error_code::unknown_server_error));
+                                     }
+
+                                     if (topic.partitions.size() == 1) {
+                                         auto& part = topic.partitions[0];
+                                         if (part.error_code != error_code{}) {
+                                             return ss::make_exception_future<
+                                               fetch_response>(partition_error(
+                                               tp, part.error_code));
+                                         }
+                                     }
+                                 }
+
+                                 return ss::make_ready_future<fetch_response>(
+                                   std::move(res));
+                             });
```

Comment:
nit: This is a stylistic choice, but:
```suggestion
                           return b->dispatch(build_request(tp));
                       })
                       .then([b, &tp](fetch_response res) {
                           if (res.data.error_code != error_code::none) {
                               return ss::make_exception_future<fetch_response>(
                                 broker_error(b->id(), res.data.error_code));
                           }

                           const auto& topics = res.data.topics;
                           if (
                             topics.size() != 1
                             || topics[0].partitions.size() != 1()) {
                               return ss::make_exception_future<fetch_response>(
                                 partition_error(
                                   tp, error_code::unknown_server_error));
                           }

                           auto& part = topics[0].partitions[0];
                           if (part.error_code != error_code::none) {
                               return ss::make_exception_future<fetch_response>(
                                 partition_error(tp, part.error_code));
                           }
                           return ss::make_ready_future<fetch_response>(
                             std::move(res));
                       });
                 })
```

Reply:
Oh, my IDE caught up, you need the nesting due to `b`. It might be worth extracting it as a function. Consumer fetch might need something quite similar.

Reply:
> Oh, my IDE caught up, you need the nesting due to `b`. It might be worth extracting it as a function. Consumer fetch might need something quite similar.

I made the style changes but I did not add anything to consumer fetch since I do not see where we would get the topic partition from for the partition_error

Reply:
I see it would be consumer::assignment but what do you compare the size of the partitions vector to?

Reply:
This PR addresses the original issue. Why don't we continue as is and address the consumer fetch stuff after?

Reply:
https://github.com/redpanda-data/redpanda/pull/7210